### PR TITLE
feat: add ConstructorEager type

### LINKED ISSUE #1013: chore: type `constructor_eager` as `ConstructorEager` instead of `Constructor`
In here

https://github.com/narwhals-dev/narwhals/blob/9353ef9f681f032feee777fabbb8db8ae98aa79d/tests/utils.py#L20

we have `Constructor`

The task is here it to:
- make `ConstructorEager`, with `IntoDataFrame` instead of `IntoFrame`
- in tests that have `constructor_eager: Any`, use instead `constructor_eager: ConstructorEager`. You may need to import it, see other tests for examples

### PR DESCRIPTION
**Pull request by lucianosrp**
## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [x] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #1013

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Comment by lucianosrp** [comment]
Getting some issues with type checking. 

```
tests/expr_and_series/sort_test.py:30: error: Item "NativeFrame" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
tests/expr_and_series/sort_test.py:30: error: Item "DataFrame[Any]" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
tests/expr_and_series/sort_test.py:30: error: Item "DataFrameLike" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
```

Should I also add  `equals` to the NativeFrame protocol?

**Comment by lucianosrp** [comment]
Not sure how to handle the mypy failure on `interchange_to_arrow_test` and `interchange_schema_test` this time.
@FBruzzesi , perhaps you could have a look? I am sure I am missing something very obvious haha

File: narwhals/typing.py [hunk 1]
```diff
@@ -28,9 +28,11 @@ class NativeFrame(Protocol):
         def columns(self) -> Any: ...
 
         def join(self, *args: Any, **kwargs: Any) -> Any: ...
+        def __getitem__(self, *args: Any, **kwargs: Any) -> Any: ...
 
     class DataFrameLike(Protocol):
         def __dataframe__(self, *args: Any, **kwargs: Any) -> Any: ...
+        def __getitem__(self, *args: Any, **kwargs: Any) -> Any: ...
```
Comment (MarcoGorelli) [inline narwhals/typing.py:36]: I don't think we should be modifying this

**Review by MarcoGorelli — COMMENTED** [review]
thanks @lucianosrp ! just a comment, but I think the rest is fine

**Comment by MarcoGorelli** [comment]
> Getting some issues with type checking.
> 
> ```
> tests/expr_and_series/sort_test.py:30: error: Item "NativeFrame" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
> tests/expr_and_series/sort_test.py:30: error: Item "DataFrame[Any]" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
> tests/expr_and_series/sort_test.py:30: error: Item "DataFrameLike" of "NativeFrame | DataFrame[Any] | DataFrameLike" has no attribute "equals"  [union-attr]
> ```
> 
> Should I also add `equals` to the NativeFrame protocol?
> 
> EDIT: I ended up removing the typehint on `sort_test.py:30` as I don't think any dataframe would have the `equals` method (?)

I think you can rewrite the test as follows:
```
    result = (
        df.select(
            "a",
            nw.col("b").sort(descending=descending, nulls_last=nulls_last),
        )
    )
    compare_dicts(result, expected)
```

**Comment by lucianosrp** [comment]
This has been stale for a while. I'll try to make it right this weekend 🔥

**Review by MarcoGorelli — APPROVED** [review]
awesome, thanks @lucianosrp !

**Pull request merged**